### PR TITLE
Don't expose methods and attrs for internal usage

### PR DESCRIPTION
**Pull request by kamipo**

### CODE REVIEW

**Comment by kaspth** [comment]
All these classes are nodoc'ed, why not keep the Git history intact and then move to private on Ruby 2.3+?

**Comment by kamipo** [comment]
I changed to keep the Git history intact as much as possible.

I think that correct visibility helps to understand the code (a reason like #26076).

The TODO comments in the previous implementation followed 21e5fd4. Probably the whole of the TODO comments will be removed and will be changed attrs visibility to private in the future.

File: activerecord/lib/active_record/reflection.rb [hunk 1]
```diff
@@ -581,14 +581,11 @@ def add_as_through(seed)
         seed + [self]
       end
 
-      protected
-
+      private
         def actual_source_reflection # FIXME: this is a horrible name
```
Comment (matthewd) [inline activerecord/lib/active_record/reflection.rb]: This isn't really private, so the change seems more misleading than explanatory.

File: activerecord/lib/active_record/associations/alias_tracker.rb [hunk 2]
```diff
@@ -5,6 +5,7 @@ module Associations
     # Keeps track of table aliases for ActiveRecord::Associations::JoinDependency
     class AliasTracker # :nodoc:
       attr_reader :aliases
+      private :aliases
```
Comment (kaspth) [inline activerecord/lib/active_record/associations/alias_tracker.rb]: This seems worse than the originally proposed `protected` to me.

**Comment by kaspth** [comment]
Regardless of what I've said in the past, I think we should just leave these be and deal with them once we're Ruby 2.3+.

@matthewd your call 👍

File: activerecord/lib/active_record/reflection.rb [hunk 3]
```diff
@@ -983,19 +983,18 @@ def add_as_through(seed)
...
+          if options[:source_type]
+            through_reflection.add_as_polymorphic_through self, a
+          else
+            through_reflection.add_as_through a
+          end
+        end
 
         def actual_source_reflection # FIXME: this is a horrible name
-          source_reflection.send(:actual_source_reflection)
+          source_reflection.actual_source_reflection
```
Comment (kamipo) [inline activerecord/lib/active_record/reflection.rb]: > This isn't really private, so the change seems more misleading than explanatory.

Good point. Actually `actual_source_reflection` is called in `ThroughReflection`.
But `ThroughReflection` don't inherite `AssociationReflection`, so I moved `actual_source_reflection` from `AssociationReflection` to `AbstractReflection`.

File: activerecord/lib/active_record/reflection.rb [hunk 4]
```diff
@@ -983,19 +983,19 @@ def add_as_through(seed)
...
+      protected
+        def actual_source_reflection # FIXME: this is a horrible name
+          source_reflection.actual_source_reflection
         end
-      end
 
       private
-
-        def actual_source_reflection # FIXME: this is a horrible name
-          source_reflection.send(:actual_source_reflection)
```
Comment (kamipo) [inline activerecord/lib/active_record/reflection.rb]: Actually `actual_source_reflection` is called in `ThroughReflection`.
But `ThroughReflection` don't inherite `AssociationReflection`, so I moved `actual_source_reflection` from `AssociationReflection` to `AbstractReflection`.

File: activerecord/lib/active_record/associations/alias_tracker.rb (hunk 2 above)
Comment (kamipo) [inline activerecord/lib/active_record/associations/alias_tracker.rb]: Okay I'll change it back to originally proposed implementation.

**Pull request merged**